app/ml_models/rnn/data_tools.py

```diff
@@ -0,0 +1,50 @@
+import torch
+import os
+from torch.utils.data import Dataset, DataLoader
+from collections import Counter
+
+
+class WordLevelDataset(Dataset):
+    def __init__(self,
+                 prefix: str = '../../../data/',
+                 filename_dataset: str = 'straattaal.txt',
+                 filename_vocab: str = 'vocabulary.txt'):
+        filename_dataset = os.path.join(prefix, filename_dataset)
+        filename_vocab = os.path.join(prefix, filename_vocab)
+
+        with open(filename_dataset, 'r', encoding='utf-8') as f:
+            lines = f.read().strip().lower()
+            self.words = [s.strip().replace('\t', '')
+                          for s in lines.split("\n")]
+        with open(filename_vocab, 'r', encoding='utf-8') as f:
+            self.vocabulary = list(f.read())
+        self.vocabulary += ['<BOS>', '<EOS>']
+        self.vocabulary_size = len(self.vocabulary)
+        self.char_to_idx_dict = {ch: i for i, ch in enumerate(self.vocabulary)}
+        self.idx_to_char_dict = {i: ch for i, ch in enumerate(self.vocabulary)}
+
+    def __len__(self):
+        return len(self.words)
+
+    def __getitem__(self, i):
+        s1 = [
+            self.char_to_idx_dict[z]
+            for z in ["<BOS>"] + list(self.words[i])
+        ]
+        s2 = [
+            self.char_to_idx_dict[z]
+            for z in list(self.words[i]) + ["<EOS>"]
+        ]
+        return torch.LongTensor(s1), torch.LongTensor(s2)
+
+    def convert_to_string(self, char_ix):
```

Comment:
Do we want char_ix to be any kind of iterable here or are we going to be more strict and "enforce" (yay python) it to be a list or any structure that you're currently using? In any case, we should probably start adding type hints as well.

Reply:
Yes, should be nicely handled by the vocabulary class in the future!